R/unnecessary_nesting_linter.R

```diff
@@ -0,0 +1,168 @@
+  )
+  exit_call_expr <- glue("
+  expr[SYMBOL_FUNCTION_CALL[{xp_text_in_table(exit_calls)}]]
+  ")
+  # block IF here for cases where a nested if/else is entirely within
+  #   one of the branches.
+  # TODO(michaelchirico): we could try and make the parallel exits requirement
+  #   more recursive, but it's a pain to do so.
+  no_signal_call_expr <- glue("
+  expr[
+    OP-LEFT-BRACE
+    and expr[
+      position() = last()
+      and not(IF)
+      and not(expr[SYMBOL_FUNCTION_CALL[{xp_text_in_table(signal_calls)}]])
+    ]
+  ]
+  ")
+  # condition for ELSE should be redundant, but include for robustness
+  # condition on parent::expr[IF] ensures we're at the first `if` of a sequence of if/else statements
+  # condition on expr uses following-sibling or preceding-sibling to ensure
+  #   that the other expr falls on a different branch (earlier used separate
+  #   conditions on two expr[], but seems to allow any branch with >1 statement
+  #   to lead to a lint.
+  # use position() = last() to ignore any expr but the last one in any branch.
+  if_else_exit_xpath <- glue("
+  //expr[
+    IF
+    and ELSE
+    and not(parent::expr[IF])
+    and expr[
+      OP-LEFT-BRACE
+      and expr[position() = last() and {exit_call_expr}]
+      and (
+        following-sibling::{no_signal_call_expr}
+        or preceding-sibling::{no_signal_call_expr}
+      )
+    ]
+  ]
+  ")
+
+  # several carve-outs of common cases where single-expression braces are OK
+  #   - control flow statements: if, for, while, repeat, switch()
+  #       + switch() is unique in being a function, not a language element
+  #       + include foreach() as a common package-based for loop extension
+  #   - function definitions
+  #       + includes purrr-like anonymous functions as ~ {...}
+  #   - rlang's double-brace expressions like {{ var }}
+  #       + NB: both braces would trigger here, so we must exclude both of them
+  #   - any expression ending like `})` or `}]`
+  #       + note that nesting is not improved by "fixing" such cases,
+  #         and could also be worsened
+  #       + motivated by the most common cases:
+  #          * test_that("test", { expr })
+  #          * with(x, { expr }) / within(x, { expr })
+  #          * suppressWarnings({ expr })
+  #          * DataTable[, { expr }]
+  #          * DataTable[, col := { expr }] <- requires carve-out for `:=`
+  unnecessary_brace_xpath <- "
+  //OP-LEFT-BRACE
+    /parent::expr[
+      count(expr) = 1
+      and not(preceding-sibling::*[
+        self::FUNCTION
+        or self::FOR
+        or self::IF
+        or self::WHILE
+        or self::REPEAT
+        or self::expr/SYMBOL_FUNCTION_CALL[text() = 'switch']
+        or self::expr/expr/SYMBOL_FUNCTION_CALL[text() = 'foreach']
+        or self::OP-TILDE
+        or self::LEFT_ASSIGN[text() = ':=']
+      ])
+      and not(expr/OP-LEFT-BRACE)
+      and not(preceding-sibling::OP-LEFT-BRACE)
+      and not(
```

Comment:
Add a similar carve-out for `OP-LEFT-BRACE`?

Reply:
do you have an example in mind? that will help for testing

Reply:
```r
tryCatch({
  main()
}, error = identity, warning = identity)
```

Reply:
this should not lint you mean? i.e. using either `({` or `})` will exempt the expression (regardless of what `brace_linter()` says)

Reply:
Yes. What does brace_linter say? I don't get lints for my example.

Reply:
Nothing, but wondering if it should (there at least seems to be a conflict with {styler})